pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java

```diff
@@ -400,117 +436,262 @@ private void rewriteRawSVForwardIndex(String column, ColumnMetadata existingColM
+            creator.putFloat(val);
+          } else {
+            float[] floats = reader.getFloatMV(i, readerContext);
+            creator.putFloatMV(floats);
+          }
+        }
+        break;
+      }
+      case DOUBLE: {
+        for (int i = 0; i < numDocs; i++) {
+          if (isSVColumn) {
+            double val = reader.getDouble(i, readerContext);
+            creator.putDouble(val);
+          } else {
+            double[] doubles = reader.getDoubleMV(i, readerContext);
+            creator.putDoubleMV(doubles);
+          }
+        }
+        break;
+      }
+      case STRING: {
+        for (int i = 0; i < numDocs; i++) {
+          if (isSVColumn) {
+            String val = reader.getString(i, readerContext);
+            creator.putString(val);
+          } else {
+            String[] strings = reader.getStringMV(i, readerContext);
+            creator.putStringMV(strings);
+          }
+        }
+        break;
+      }
+      case BYTES: {
+        for (int i = 0; i < numDocs; i++) {
+          if (isSVColumn) {
+            byte[] val = reader.getBytes(i, readerContext);
+            creator.putBytes(val);
+          } else {
+            byte[][] bytesArray = reader.getBytesMV(i, readerContext);
+            creator.putBytesMV(bytesArray);
+          }
+        }
+        break;
+      }
+      case BIG_DECIMAL: {
+        Preconditions.checkState(isSVColumn, "BigDecimal is not supported for MV columns");
+        for (int i = 0; i < numDocs; i++) {
+          BigDecimal val = reader.getBigDecimal(i, readerContext);
+          creator.putBigDecimal(val);
+        }
+        break;
+      }
+      default:
+        throw new IllegalStateException("Unsupported storedType=" + reader.getStoredType() + " for column=" + column);
+    }
+  }
+
+  private void forwardIndexReadDictWriteRawHelper(String column, ColumnMetadata existingColumnMetadata,
+      ForwardIndexReader reader, ForwardIndexCreator creator, int numDocs, Dictionary dictionaryReader) {
+    ForwardIndexReaderContext readerContext = reader.createContext();
+    boolean isSVColumn = reader.isSingleValue();
+    int maxNumValuesPerEntry = existingColumnMetadata.getMaxNumberOfMultiValues();
+    PinotSegmentColumnReader columnReader =
+        new PinotSegmentColumnReader(reader, dictionaryReader, null, maxNumValuesPerEntry);
+    FieldSpec.DataType storedType = dictionaryReader.getValueType().getStoredType();
+
+    switch (storedType) {
+      case INT: {
+        for (int i = 0; i < numDocs; i++) {
+          if (isSVColumn) {
+            Object obj = columnReader.getValue(i);
+            int val = (int) obj;
+            creator.putInt(val);
+          } else {
+            int[] dictIds = reader.getDictIdMV(i, readerContext);
+            int[] ints = new int[dictIds.length];
+            for (int j = 0; j < dictIds.length; j++) {
+              ints[j] = (int) dictionaryReader.get(dictIds[j]);
```

Comment:
May be for SV also we can follow the same reading patter to be consistent ?

Read dictID from fwd index reader and then read the dict to get corresponding raw value and then insert.

Basically columnReader isn't technically needed ?